ff/serving/provider/postgres.go

```diff
@@ -0,0 +1,501 @@
+		"SELECT column_name FROM information_schema.columns WHERE table_name = $1 order by ordinal_position",
+		trainingSetName)
+	defer rows.Close()
+	if err != nil {
+		return nil, err
+	}
+	features := make([]string, 0)
+	for rows.Next() {
+		var column string
+		if err := rows.Scan(&column); err != nil {
+			return nil, err
+		}
+		if column != "label" {
+			features = append(features, sanitize(column))
+		}
+	}
+	columns := strings.Join(features[:], ", ")
+
+	trainingSetQry := fmt.Sprintf("SELECT %s FROM %s", columns, sanitize(trainingSetName))
+
+	rows, err = store.conn.Query(context.Background(), trainingSetQry)
+	defer rows.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	trainingData := make([]trainingRow, 0)
+	for rows.Next() {
+
+		var label interface{}
+		values, err := rows.Values()
+		if err != nil {
+			return nil, err
+		}
+		// Clean this up more
+		featureVals := make([]interface{}, len(values)-1)
+		for i, value := range values {
+			if value == nil {
+				featureVals[i] = value
+			} else if i < len(values)-1 {
+				v := value.(map[string]interface{})
+				item := postgresTableItem{
+					Value:    v["value"],
+					ItemType: v["type"].(string),
+				}
+				featureVals[i] = castTableItemType(item)
+			} else {
+				v := value.(map[string]interface{})
+				item := postgresTableItem{
+					Value:    v["value"],
+					ItemType: v["type"].(string),
+				}
+				label = castTableItemType(item)
+			}
+		}
+		trainingData = append(trainingData, trainingRow{
+			Features: featureVals,
+			Label:    label,
+		})
+	}
+	return newTrainingRowsIterator(trainingData), nil
+}
+
+type postgresOfflineTable struct {
+	conn *pgxpool.Pool
+	ctx  context.Context
+	name string
+}
+
+func newPostgresOfflineTable(conn *pgxpool.Pool, name string) (*postgresOfflineTable, error) {
+	tableCreateQry := fmt.Sprintf("CREATE TABLE %s (entity VARCHAR, value JSONB, ts timestamptz)", sanitize(name))
+	_, err := conn.Exec(context.Background(), tableCreateQry)
+	if err != nil {
+		return nil, err
+	}
+	return &postgresOfflineTable{
+		conn: conn,
+		name: name,
+	}, nil
+}
+
+func (table *postgresOfflineTable) serialize(v interface{}) ([]byte, error) {
+	item := postgresTableItem{
+		Value:    v,
+		ItemType: fmt.Sprintf("%T", v),
+	}
+	return json.Marshal(item)
+}
+
+func (table *postgresOfflineTable) deserialize(v []byte) (interface{}, error) {
+	item := postgresTableItem{}
+	if err := json.Unmarshal(v, &item); err != nil {
+		return nil, err
+	}
+	return item.Value, nil
+}
+
+// Check logic on this one
+func (table *postgresOfflineTable) Write(rec ResourceRecord) error {
+	tb := sanitize(table.name)
+
+	if err := rec.check(); err != nil {
+		return err
+	}
+
+	value, err := table.serialize(rec.Value)
+	if err != nil {
+		return err
+	}
+	if exists, err := table.resourceExists(rec); err != nil {
+		return err
+	} else if !exists {
+		query := fmt.Sprintf("INSERT INTO %s (entity, value, ts) VALUES ($1, $2, $3) RETURNING entity, value, ts", tb)
```

Comment:
UPSERT is a thing in postgres and snowflake now I think

Reply:
Yep, INSERT ... ON CONFLICT